### PR TITLE
Reenable filesystem for ios builds

### PR DESCRIPTION
**Pull request by kikaxa**
As per comment https://github.com/opencv/opencv/pull/20010#commitcomment-51027021

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/include/opencv2/core/utils/filesystem.private.hpp [hunk 1]
```diff
@@ -16,8 +16,8 @@
 #      define OPENCV_HAVE_FILESYSTEM_SUPPORT 1
 #  elif defined(__APPLE__)
 #    include <TargetConditionals.h>
-#    if (defined(TARGET_OS_OSX) && TARGET_OS_OSX) || (!defined(TARGET_OS_OSX) && !TARGET_OS_IPHONE)
-#      define OPENCV_HAVE_FILESYSTEM_SUPPORT 1 // OSX only
+#    if (defined(TARGET_OS_OSX) && TARGET_OS_OSX) || (defined(TARGET_OS_IOS) && TARGET_OS_IOS)
+#      define OPENCV_HAVE_FILESYSTEM_SUPPORT 1 // OSX, iOS only
```
Comment (alalek) [inline modules/core/include/opencv2/core/utils/filesystem.private.hpp:20]: This condition is from 3.4 branch: https://github.com/opencv/opencv/blame/3.4/modules/core/include/opencv2/core/utils/filesystem.private.hpp#L19

relates #10018  (before any of 4.x releases)

---

What functionality you want to use on iOS?

Comment (kikaxa) [reply modules/core/include/opencv2/core/utils/filesystem.private.hpp:20]: We found ::fs:: , and (strangely) some parts of videoio not working.

Opencv builds and works fine in our (limited) test cases for ios, macos when OPENCV_HAVE_FILESYSTEM_SUPPORT=1.

Comment (alalek) [reply modules/core/include/opencv2/core/utils/filesystem.private.hpp:20]: > `::fs::`

This API should be considered as internal.
There is no goal to support FS utilities through Computer Vision Library.
Some of these utilities are available through C++17 filesystem features, consider using them instead: https://en.cppreference.com/w/cpp/filesystem

---

> some parts of videoio not working

Do you mean a problem similar to #19910?

Comment (kikaxa) [reply modules/core/include/opencv2/core/utils/filesystem.private.hpp:20]: > `::fs::` This API should be considered as internal.
I agree. I have checked and we do not use ::fs on ios.
As a note, the whole namespace is listed and documented in the documentation, without any mentions of being internal.

We use cv::glob() (core/utils.hpp included by default with the core module) extensively, i have checked the  [PR](https://github.com/opencv/opencv/pull/20010/files#diff-75b7ddefa1f6e5cf61410a883f56c960ab9771bff45e3a1a0f2f8d225615b559) and it modifies this function as well

>videoio
No, we had an explicit assert/error mentioning the filesystem being disabled. The only place i found skimming through the module is loadPlugin()/getPluginCandidates()
We call cv::VideoCapture/open()

**Comment by alalek** [comment]
Rebased on 3.4 branch

---

> disabled in #20010

this note is removed as incorrect

**Review by alalek — APPROVED** [review]

**Pull request merged**